cmd/download.go

```diff
@@ -264,6 +197,119 @@ func (d download) needsSlugWhenGivenTrackOrTeam() error {
 	return nil
 }
 
+func (d download) metadata() *ws.ExerciseMetadata {
+	return &ws.ExerciseMetadata{
+		AutoApprove:  d.Solution.Exercise.AutoApprove,
+		Track:        d.Solution.Exercise.Track.ID,
+		Team:         d.Solution.Team.Slug,
+		ExerciseSlug: d.Solution.Exercise.ID,
+		ID:           d.Solution.ID,
+		URL:          d.Solution.URL,
+		Handle:       d.Solution.User.Handle,
+		IsRequester:  d.Solution.User.IsRequester,
+	}
+}
+
+func (d download) exercise() ws.Exercise {
+	return d.metadata().Exercise(d.workspace)
+}
+
+// requestFile requests a Solution file from the API, returning an HTTP response.
+// 0 Content-Length responses are swallowed, returning nil.
+func (d download) requestFile(filename string) (*http.Response, error) {
+	parsedURL, err := netURL.ParseRequestURI(
+		fmt.Sprintf("%s%s", d.Solution.FileDownloadBaseURL, filename))
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := api.NewClient(d.token, d.apibaseurl)
+	req, err := client.NewRequest("GET", parsedURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return nil, decodedAPIError(res)
+	}
+	// Don't bother with empty files.
+	if res.Header.Get("Content-Length") == "0" {
+		return nil, nil
+	}
+
+	return res, nil
+}
+
+// downloadWriter writes download contents to the file system.
+type downloadWriter struct {
+	download *download
+}
+
+// newDownloadWriter creates a downloadWriter.
+func newDownloadWriter(dl *download) *downloadWriter {
```

Comment:
I think we could get a way with just initializing a literal downloaderWriter in `runDownload` and dropping the constructor seeing as it is not doing anything special. With that being said, is there a reason for returning a pointer to a downloadWriter?